torch/csrc/profiler/python/init.cpp

```diff
@@ -326,5 +330,165 @@ void initPythonBindings(PyObject* module) {
+ * annotations. Part of the reason it is slow is because the calls go through
+ * the dispatcher, in order to make the record_function calls work with
+ * torchscript.
+ *
+ * This implementation is less safe; it doesn't work with torchscript.
+ *
+ * It implements a context manager in C++, which reduces the overhead compared
+ * to using a python context manager that calls into C++.
+ */
+
+namespace {
+struct RecordFunctionFast {
+  PyObject_HEAD PyObject* name;
+  std::unique_ptr<at::RecordFunction> guard;
+};
+
+PyObject* RecordFunctionFast_new(
+    PyTypeObject* subtype,
+    PyObject* args,
+    PyObject* kwargs) {
+  RecordFunctionFast* self = (RecordFunctionFast*)subtype->tp_alloc(subtype, 0);
+  if (self != nullptr) {
+    self->name = nullptr;
+    self->guard.reset();
+  }
+  return (PyObject*)self;
+}
+
+int RecordFunctionFast_init(
+    PyObject* selfGeneric,
+    PyObject* args,
+    PyObject* kwargs) {
+  auto self = (RecordFunctionFast*)selfGeneric;
+  constexpr const char* kwlist[] = {"name", nullptr};
+  PyObject* name = nullptr;
+  if (!PyArg_ParseTupleAndKeywords(
+          args, kwargs, "O", const_cast<char**>(kwlist), &name)) {
+    return -1;
+  }
+  if (name) {
+    Py_INCREF(name);
+    self->name = name;
+  }
+  return 0;
+}
+
+void RecordFunctionFast_dealloc(PyObject* selfGeneric) {
+  auto self = (RecordFunctionFast*)selfGeneric;
+  if (self->name) {
+    Py_DECREF(self->name);
+  }
+  if (self->guard) {
+    self->guard.reset();
+  }
+  Py_TYPE(self)->tp_free(self);
+}
+
+PyObject* RecordFunctionFast_enter(PyObject* selfGeneric, PyObject* unused) {
+  HANDLE_TH_ERRORS
+  if (torch::profiler::impl::ProfilerStateBase::get() != nullptr) {
+    auto self = (RecordFunctionFast*)selfGeneric;
+    TORCH_INTERNAL_ASSERT(
+        !self->guard,
+        "Trying to enter a new record_function_fast context but the guard is unexpectedly already set");
+    TORCH_INTERNAL_ASSERT(
+        self->name,
+        "Trying to enter a new record_function_fast context but self->name is not set");
+    self->guard =
+        std::make_unique<at::RecordFunction>(at::RecordScope::FUNCTION);
+    THPUtils_checkString(self->name);
+    self->guard->before(THPUtils_unpackString(self->name));
+  }
+  Py_RETURN_NONE;
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject* RecordFunctionFast_exit(PyObject* selfGeneric, PyObject* unused) {
+  HANDLE_TH_ERRORS
+  if (torch::profiler::impl::ProfilerStateBase::get() != nullptr) {
+    auto self = (RecordFunctionFast*)selfGeneric;
+    TORCH_INTERNAL_ASSERT(
+        self->guard,
+        "Trying to exit an active record_function_fast context but no guard is set");
+    self->guard.reset();
+  }
+  Py_RETURN_NONE;
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject* RecordFunctionFast_set_name(PyObject* selfGeneric, PyObject* pyName) {
+  HANDLE_TH_ERRORS
+  auto self = (RecordFunctionFast*)selfGeneric;
+  if (self->name) {
+    Py_DECREF(self->name);
+    self->name = nullptr;
+  }
```

Comment:
```suggestion
  Py_Clear(self->name);
```

The current code is not safe as self->name is in a bad state between these two lines (keep in mind that decref can deallocate the object and end up running arbitrary code)